### PR TITLE
Add bounds support for geogrid agg on shapes

### PR DESCRIPTION
**Pull request by talevy**
This commit adds support for `geo-grid` aggregations with `bounds` parameter
on geo_shape doc values

This also modifies the existing geo_point handling of `bounds` to be consistent with shapes.

### CODE REVIEW

**Comment by talevy** [comment]
This PR is still a WIP because its tests exposed existing bugs in the TriangleTreeReader#relate logic.  This logic is well tested in `TriangleTreeTests#testRectangleShape`, but the assertions there are not necessarily what we want when evaluating tiles.

The picture below depicts the unwanted behavior:

Screen Shot 2020-02-05 at 4 31 28 PM

These edge tiles should not be accounted for, outer edges in the triangle tree that are collinear with the queried tiles should not be considered CROSSES in these cases if these are not to be included.

**Comment by talevy** [comment]
run elasticsearch-ci/packaging-sample-matrix-unix

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoGridTiler.java [hunk 1]
```diff
@@ -147,8 +150,9 @@ private int setValuesForFullyContainedTile(String hash, GeoShapeCellValues value
         }
     }
 
-    class GeoTileGridTiler implements GeoGridTiler {
-        public static final GeoTileGridTiler INSTANCE = new GeoTileGridTiler();
+    class GeoTileGridTiler<G extends MultiGeoValues.GeoValue> implements GeoGridTiler<G> {
+        public static final GeoTileGridTiler INSTANCE = new GeoTileGridTiler<>();
+        public static final GeoTileGridTiler BOUNDED_INSTANCE = new GeoTileGridTiler<BoundedGeoShapeCellValues.BoundedGeoValue>();
```
Comment (iverase) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoGridTiler.java]: I guess I am missing something but I do not understand the benefits of having this two instances?

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoGridTiler.java [hunk 2]
```diff
@@ -259,8 +268,8 @@ protected int setValuesByRasterization(int xTile, int yTile, int zTile, GeoShape
             return valuesIndex;
         }
 
-        private int setValuesForFullyContainedTile(int xTile, int yTile, int zTile,
-                                                   GeoShapeCellValues values, int valuesIndex, int targetPrecision) {
+        private int setValuesForFullyContainedTile(int xTile, int yTile, int zTile, CellValues values, int valuesIndex,
+                                                   int targetPrecision) {
```
Comment (iverase) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoGridTiler.java]: Don't we need to filter out  the tiles that do not intersect with the provided bounds (if provided)

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoGridTiler.java [hunk 3]
```diff
@@ -189,17 +193,21 @@ public int setValues(GeoShapeCellValues values, MultiGeoValues.GeoValue geoValue
...
             int minXTile = GeoTileUtils.getXTile(bounds.minX(), (long) tiles);
             int minYTile = GeoTileUtils.getYTile(bounds.maxY(), (long) tiles);
             int maxXTile = GeoTileUtils.getXTile(bounds.maxX(), (long) tiles);
             int maxYTile = GeoTileUtils.getYTile(bounds.minY(), (long) tiles);
             int count = (maxXTile - minXTile + 1) * (maxYTile - minYTile + 1);
             if (count == 1) {
-                values.resizeCell(1);
-                values.add(0, GeoTileUtils.longEncodeTiles(precision, minXTile, minYTile));
-                return 1;
+                Rectangle rectangle = GeoTileUtils.toBoundingBox(minXTile, minYTile, precision);
```
Comment (iverase) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoGridTiler.java]: It is unfortunate that we need to build the rectangle evening the unbounded case

File: server/src/main/java/org/elasticsearch/index/fielddata/MultiGeoValues.java [hunk 4]
```diff
@@ -58,7 +59,7 @@
  * The set of values associated with a document might contain duplicates and
  * comes in a non-specified order.
  */
-public abstract class MultiGeoValues {
+public abstract class MultiGeoValues <G extends MultiGeoValues.GeoValue> {
```
Comment (iverase) [inline server/src/main/java/org/elasticsearch/index/fielddata/MultiGeoValues.java]: I think the javadocs do not reflect the way the object is iterated?

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoGridTiler.java [hunk 5]
```diff
@@ -189,17 +193,21 @@ public int setValues(GeoShapeCellValues values, MultiGeoValues.GeoValue geoValue
...
+                GeoRelation relation = geoValue.relate(rectangle);
+                if (relation != GeoRelation.QUERY_DISJOINT) {
```
Comment (iverase) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoGridTiler.java]: It is unfortunate that we need to build the rectangle even in the Unbounded case.

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoGridTiler.java (hunk 2 above)
Comment (iverase) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoGridTiler.java]: I think we need to filter out here tiles that do not intersect with the provided bounds (if provided)

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoGridTiler.java (hunk 1 above)
Comment (iverase) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoGridTiler.java]: I guess I am missing something but I cannot tell the benefits of having two instances, they seem to be used the same way?

**Review by iverase — CHANGES_REQUESTED** [review]
I think the biggest issue at the moment in the PR is that we need to filter tiles when we collect them from fully contained tiles. This makes me wonder that this should be different between the bounded and unbounded case, what do you think?

**Comment by iverase** [comment]
I am not sure if the current approach is the most efficient. We are wrapping the bounds around the geo value, I think it would be more efficient to wrap it around the GridTiler? the tiler can call an abstract method to check if the tile is in bounds. For the unbounded case the answer will always be true, for the bounder, it will perform the check against the provided bounds.

**Comment by talevy** [comment]
I am still thinking of how to best reduce the duplicate code around the GeoBoundingBox parts of both Geohash and Geotile, but let me know if the update is what you had in mind! thanks Ignacio

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/BoundedGeoTileGridTiler.java [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+            boundsEastRight = geoBoundingBox.right();
+            boundsWestLeft = 0;
+            boundsWestRight = 0;
+            crossesDateline = false;
+        }
+    }
+
+    public int advancePointValue(long[] values, double x, double y, int precision, int valuesIdx) {
+        long hash = encode(x, y, precision);
+        if (cellIntersectsGeoBoundingBox(GeoTileUtils.toBoundingBox(hash))) {
```
Comment (iverase) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/BoundedGeoTileGridTiler.java]: I think in the case of a point, we can just check that the point is in bounds? We save having to compute the rectangle.

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/BoundedGeoTileGridTiler.java [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+    boolean cellIntersectsGeoBoundingBox(Rectangle rectangle) {
+        return (boundsTop >= rectangle.getMinY() && boundsBottom <= rectangle.getMaxY()
+            && (boundsEastLeft <= rectangle.getMaxX() && boundsEastRight >= rectangle.getMinX()
+            || (crossesDateline && boundsWestLeft <= rectangle.getMaxX() && boundsWestRight >= rectangle.getMinX())));
+    }
+
+    @Override
+    public GeoRelation relateTile(MultiGeoValues.GeoValue geoValue, int xTile, int yTile, int precision) {
+        Rectangle rectangle = GeoTileUtils.toBoundingBox(xTile, yTile, precision);
+        if (cellIntersectsGeoBoundingBox(rectangle)) {
```
Comment (iverase) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/BoundedGeoTileGridTiler.java]: +1

**Comment by iverase** [comment]
@tal, yes, I think in this way the unbounded case behaves the same as before. Thanks!

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/BoundedGeoTileGridTiler.java (hunk 6 above)
Comment (talevy) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/BoundedGeoTileGridTiler.java]: I believe we discussed this and decided we would like to keep a consistent definition between points and shapes where we only filter which cells we consider with regards to the geoBoundingBox. this means the point need not be within the geoBoundingBox

Comment (iverase) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/BoundedGeoTileGridTiler.java]: yes, yes, my bad :(

Comment (iverase) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/BoundedGeoTileGridTiler.java]: It is exactly how we spoke and I think it should be

Comment (talevy) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/BoundedGeoTileGridTiler.java]: 👍

**Comment by talevy** [comment]
I am not sure there is a much better way to re-use the bounded logic that is less verbose. I think it may only be worth re-visiting once an additional geogrid agg is every introduced.

so this PR is ready for a final pass if you'd like. thanks @iverase!

**Review by iverase — APPROVED** [review]
LGTM

**Comment by talevy** [comment]
thanks!

**Pull request merged**